### PR TITLE
[pvr] Radio RDS support

### PR DESCRIPTION
**Pull request by AlwinEsch**
This changes implement handling about RDS signal from Radio. The related data need to be send with other streams from add-on and become handled inside DVDPlayer.

Data become then translated and stored in own file Item information tag and can be used from skin and other add-ons over GUIInfoManager. Also if Radiotext plus is present, the Music Artist and Title on MusicInfoTag become set. Further is for Radiotext+ a new dialog present. For this are also 3 background images (currently only 3, it gives 32 genres on RDS).

The normal Radiotext become shown on PVR channels window. For this is on skin a setting to show only EPG or RDS and if setting there enabled are both on two places

Further support this the RaSS images send from RDS (currently only German SWR).

If on settings enabled the traffic messages can be passed outside. Allow to create a add-on which can handle related data.

The needed data format is Universal Encoder Communication Protocol (UECP SPB 490).

Have created a new add-on which use the library from RTL-SDR (http://sdr.osmocom.org/trac/wiki/rtl-sdr) to handle FM Radio with RDS and also created changes on VDR VNSI addon and related source plugin.

Findable currently under following links:
- https://github.com/AlwinEsch/pvr.rtl.radiofm
- https://github.com/AlwinEsch/kodi-pvr-addons/tree/vnsi-rds-support
- https://github.com/AlwinEsch/vdr-plugin-vnsiserver/tree/vnsi-rds-support

Also are on Patch 6 and 7 changes not direct related to this request. 
The first is a hack to open add-on channel settings dialogues (I don't like this way!), searched also with the ADSP about the problem. With a for me unknown reason a add-on dialogue becomes always opened behind KODI's one.
The second allow the usage from PVR related buttons over the Music OSD dialogue.

Here are some screen shots:

PVR radio window
screenshot1

Full screen view with add-on signal
screenshot2

RaSS information image and also enabled for EPG (currently I have no channel with EPG thats why empty)
screenshot5

Dialogue which can show the information from RT+ (most present until now under classical radio).
screenshot6

PVR settings
screenshot4

Settings on skin
screenshot3

### CODE REVIEW

**Comment by AlwinEsch** [comment]
Hello all,

can someone help me to add windows and other OS Build files?

Thanks a lot

**Comment by Jalle19** [comment]
@AlwinEsch I can do the Windows changes, I've been meaning to test this as soon as I have time (have a spare RTL2832 stick that I haven't found any use for until now).

**Comment by AlwinEsch** [comment]
@Jalle19 thanks

**Comment by Jalle19** [comment]
Ping @ronie for the skinning parts

**Comment by Jalle19** [comment]
I take it the PVR addon can only be used on Linux, and the backend cannot be separated from the client? I'm trying to figure out how to test this since I don't have an antenna jack near my HTPC.

**Comment by opdenkamp** [comment]
only did a quick sanity check of the files changed, not the new ones.

@Jalle19 why would this only work on a local backend on linux? it's just stream data?

**Comment by Jalle19** [comment]
@opdenkamp I was actually referring to the addon he made.

**Comment by AlwinEsch** [comment]
@Jalle19 has seen that the needed library is also available for windows (http://sdr.osmocom.org/trac/attachment/wiki/rtl-sdr/RelWithDebInfo.zip), thats why i thought is also compatible. Has set in cmake file only a compiler improve for gcc but this system is new for me and no idea how this is handled in windows.

@opdenkamp thanks for the comments, I perform currently a cleanup.

**Comment by Jalle19** [comment]
@AlwinEsch cool, didn't know there was a Windows library for it. I'd love to be able to have the RTL stick plugged into the server and use the addon on any client, but it seems that's not possible since it interfaces directly with the hardware.

**Comment by AlwinEsch** [comment]
@opdenkamp have cleaned up based upon your comments and changed the RDS info tag to a shared_ptr and only one present on related stream class. Is nice.

Is also the Music Info tag planned to use this way? Currently the Artist and Title on this becomes set from RDS stream if present (RT+).

**Comment by ronie** [comment]
i would drop the genre specific backgrounds from confluence.
adding 32 fullscreen background images would take up too much space,
and only including a small subset (3, as you did) does not make much sense.

**Comment by AlwinEsch** [comment]
Hello @ronie thanks for the comments, I do a cleanup in the next days.

With the backgrounds was from me to have the dialog not always complete empty. If a radio is heard a longer time and the radio send a lot of infos it can be complete full, but most time it is only this. With 3 was my idea only to know how it looks and to add maybe later more.
But is right, takes a lot of space and I remove it then.

**Comment by ronie** [comment]
thanx for the fixes! :-)

one more thing i forgot to ask, why did you add a complete new window to the PVR section (MyPVRRadioChannels.xml)?

can't you just add the new view to MyPVRChannels.xml?

**Comment by AlwinEsch** [comment]
Thought with it to reduce amount of work for TV channels, the most people look TV and in Radio are a lot of things which becomes then never used. The other one was to allow other skins to make own view for radio.
It can be combined, what think you?

**Comment by ronie** [comment]
> The other one was to allow other skins to make own view for radio.

i think skins can already create separate views for Radio by using the 'IsRadio' window property ?

we should try to avoid adding new windows as much as possible.
adding a new window + view is way more work than adding an additional view to an existing window.
the pvr section already requires an insane amount of windows/dialogs (compared to other sections in kodi)
and skinners won't be too thrilled having to add yet another window for it.

**Comment by opdenkamp** [comment]
any updates on this?

**Comment by AlwinEsch** [comment]
Has only updated few days ago with master, waiting about other answers on Team Forum. Is about Rass support inside, the Radio channel with it stopped this, few months ago. And currently no radio channels with it present. If not come a another way to use it then I must remove it.

**Comment by mkortstiege** [comment]
IMO it's overkill support something for just one channel that even dropped it. Please remove if there are no good reasons to keep and maintain it.

**Comment by AlwinEsch** [comment]
Thanks @mkortstiege, is see this just the same.

**Comment by opdenkamp** [comment]
Agreed

**Comment by AlwinEsch** [comment]
What is the status about this? There are currently no API changes more needed on KODI (was brought in earlier).

**Comment by FernetMenta** [comment]
@opdenkamp mind having a look at this again?

**Comment by opdenkamp** [comment]
will do. and i still have some crash to fix related to add-ons in the 
repos too :)

**Comment by AlwinEsch** [comment]
jenkins build this please

**Comment by AlwinEsch** [comment]
Can someone help me please to fix compile faults on win and OSX?

**Comment by Memphiz** [comment]
And again - why are people triggering jenkins when its obvious that there need to be project adaptions first? Its a waste of resources - especially on days like this when jenkins is pretty much crowded already ...

**Comment by Memphiz** [comment]
project adaptions incoming ...

**Comment by AlwinEsch** [comment]
Before it was OK and has want to only a recheck of him.

**Comment by Paxxi** [comment]
@Memphiz are you doing VS as well?

**Comment by AlwinEsch** [comment]
Is updated and your parts @Memphiz are included.

**Comment by Memphiz** [comment]
jenkins build this please (if win32 doesn't work because of missing project adjustments then its because i edited it only with text editor - no windows here ...)

**Comment by AlwinEsch** [comment]
Is updated with current xbmc branch and the numbers removed.

**Comment by AlwinEsch** [comment]
Pull request is updated to the latest kodi version and tested, seems to be still ok.

jenkins build this please

File: xbmc/pvr/dialogs/GUIDialogPVRRadioRDSInfo.cpp [hunk 1]
```diff
@@ -0,0 +1,416 @@
...
+#include "messaging/ApplicationMessenger.h"
+#include "pvr/channels/PVRRadioRDSInfoTag.h"
+
+using namespace PVR;
+
+#define CONTROL_BTN_OK    10
+#define SPIN_CONTROL_INFO 21
+#define TEXT_INFO         22
+#define CONTROL_NEXT_PAGE 60
+#define CONTROL_INFO_LIST 70
```
Comment (AlwinEsch) [inline xbmc/pvr/dialogs/GUIDialogPVRRadioRDSInfo.cpp]: Hi @phil65 about your question with "Radio text info list" skin group, on skin was a mistake from me during testing and the id was temporary removed. Is normally 70 and becomes set visible on the line 108 if availabe.

File: xbmc/GUIInfoManager.cpp [hunk 2]
```diff
@@ -3767,6 +3904,174 @@ std::string CGUIInfoManager::GetPlaylistLabel(int item, int playlistid /* = PLAY
...
+
+  case RDS_CONDUCTOR:
+    return tag.GetConductor();
+
+  case RDS_ALBUM:
+    return tag.GetAlbum();
+
+  case RDS_ALBUM_TRACKNUMBER:
+    {
+      if (!tag.GetAlbumTrackNumber() > 0)
```
Comment (mkortstiege) [inline xbmc/GUIInfoManager.cpp]: xcode suggests to add parentheses

File: xbmc/cores/dvdplayer/DVDPlayerRadioRDS.cpp [hunk 3]
```diff
@@ -0,0 +1,1748 @@
...
+  }
+
+  return text;
+}
+
+unsigned short crc16_ccitt(const unsigned char *daten, int len, bool skipfirst)
+{
+  // CRC16-CCITT: x^16 + x^12 + x^5 + 1
+  // with start 0xffff and result invers
+  register unsigned short crc = 0xffff;
```
Comment (mkortstiege) [inline xbmc/cores/dvdplayer/DVDPlayerRadioRDS.cpp]: register is deprecated and so causing a compile warning.

File: xbmc/GUIInfoManager.cpp (hunk 2 above)
Comment (AlwinEsch) [inline xbmc/GUIInfoManager.cpp]: Better is that I remove the "!" is a mistake.

**Comment by bam80** [comment]
Does it work for DVB-T2 radio channels?

**Pull request merged**